File: llama_mobile-ios-SDK/Tests/LlamaMobileTests/LlamaEngineTests.swift
import XCTest
import LlamaMobile
import llama_mobile

/// Model fixture paths (mirrors the C++ core suite fixtures).
enum TestPaths {
    static let rootPath = "/Users/shileipeng/Documents/mygithub/llama_mobile/models"
    static let modelPath = rootPath + "/SmolLM-360M-Instruct.Q6_K.gguf"
    static let ttsModelPath = rootPath + "/OuteTTS-0.2-500M-Q6_K.gguf"
    static let altTTSModelPath = rootPath + "/Qwen3-1.7B-Multilingual-TTS.Q5_K_M.gguf"
    static let vocoderPath = rootPath + "/WavTokenizer-Large-75-F16.gguf"
    static let embeddingPath = rootPath + "/embedding/Qwen3-Embedding-0.6B-Q8_0.gguf"
    static let mmprojPath = rootPath + "/mmproj-SmolVLM-256M-Instruct-Q8_0.gguf"
    static let loraPath = rootPath + "/lora/fine-tuned-smolLM2-360M-with-LoRA-on-camel-ai-physics-f16.gguf"
    static let imageModelPath = rootPath + "/SmolVLM-256M-Instruct-Q8_0.gguf"
    static let imagePath = rootPath + "/img/image.jpg"
}

/// Conformance tests for the v2 LlamaEngine wrapper.
///
/// Threading contract under test (docs/api-contract-v2.md §8):
///   * sync forms block the calling thread — the tests therefore run blocking
///     calls inside Task.detached so the test runner is never blocked;
///   * async forms never block the caller;
///   * streaming reports `.started(requestId)` before the first token and
///     `.done` after the request finishes;
///   * `abort(requestId:)` is thread-safe and the aborted generation reports
///     stopReason == .aborted.
///
/// Model-backed tests reuse the same fixtures as the C++ core suites and skip
/// (XCTSkip) when the file is not reachable, so the suite also runs on CI/dev
/// machines without the big models. On a device, put a GGUF where
/// LLAMA_MOBILE_TEST_MODEL points, or pass it via environment.
final class LlamaEngineTests: XCTestCase {

    private var modelPath: String {
        if let env = ProcessInfo.processInfo.environment["LLAMA_MOBILE_TEST_MODEL"],
           FileManager.default.fileExists(atPath: env) {
            return env
        }
        return TestPaths.modelPath
    }

    private func requireModel() throws {
        try XCTSkipUnless(FileManager.default.fileExists(atPath: modelPath),
                          "model not available: \(modelPath)")
    }

    private func makeConfig() -> LlamaModelConfig {
        var c = LlamaModelConfig(modelPath: modelPath)
        c.engine = .cpu
        c.nCtx = 2048
        return c
    }

    // MARK: - No-model tests (pure API surface)

    func testVersionApiIsV2() throws {
        guard let ver = llama_mobile_version() else {
            return XCTFail("llama_mobile_version() returned nil")
        }
        let v = ver.pointee
        XCTAssertEqual(v.api_version, 2)
        XCTAssertNotNil(String(cString: v.string).isEmpty ? nil : v.string)
    }

    func testErrorMappingFromStatus() {
        XCTAssertEqual(LlamaError(status: LLAMA_MOBILE_OK), .generationFailed, "OK is not an error")
        XCTAssertEqual(LlamaError(status: LLAMA_MOBILE_ERR_INVALID_ARGUMENT), .invalidArgument)
        XCTAssertEqual(LlamaError(status: LLAMA_MOBILE_ERR_ALREADY_RUNNING), .alreadyRunning)
        XCTAssertEqual(LlamaError(status: LLAMA_MOBILE_ERR_ABORTED), .aborted)
        XCTAssertEqual(LlamaError(status: LLAMA_MOBILE_ERR_NOT_INITIALIZED), .notInitialized)
        XCTAssertEqual(LlamaError(status: LLAMA_MOBILE_ERR_UNSUPPORTED), .unsupported)
    }

    func testGenerateAfterCloseThrowsNotInitialized() async throws {
        try requireModel()
        let engine = try await LlamaEngine.open(makeConfig())
        engine.close()
        do {
            _ = try await engine.generate(LlamaGenerationRequest(prompt: "hi"))
            XCTFail("expected notInitialized after close")
        } catch let e as LlamaError {
            XCTAssertEqual(e, .notInitialized)
        } catch {
            XCTFail("unexpected error \(error)")
        }
    }

    func testOpenMissingModelThrows() async throws {
        var c = LlamaModelConfig(modelPath: "/nonexistent/model.gguf")
        c.engine = .cpu
        do {
            _ = try await LlamaEngine.open(c)
            XCTFail("expected open to throw for a missing model")
        } catch let e as LlamaError {
            XCTAssertTrue(e == .modelLoadFailed || e == .modelNotFound || e == .io,
                          "got \(e)")
        } catch {
            XCTFail("unexpected error \(error)")
        }
    }

    // MARK: - Model-backed tests (§8 sync/async/stream/abort)

    func testSyncOpenAndGenerate() async throws {
        try requireModel()
        let engine = try await LlamaEngine.open(makeConfig())
        defer { engine.close() }
        XCTAssertTrue(engine.isOpen)

        var req = LlamaGenerationRequest(prompt: "Complete: the capital of France is")
        req.maxTokens = 16
        req.sampling.temperature = 0
        req.sampling.topP = 1.0
        req.sampling.minP = 0
        // Sync form blocks: never call it on the actor, so hop off.
        let r = try await Task.detached { try engine.generate(req) }.value
        XCTAssertFalse(r.text.isEmpty)
        XCTAssertEqual(r.stopReason, .eos)
        XCTAssertGreaterThan(r.usage.generatedTokens, 0)
        XCTAssertGreaterThanOrEqual(r.usage.promptTokens, 0)
    }

    func testAsyncChatGenerate() async throws {
        try requireModel()
        let engine = try await LlamaEngine.open(makeConfig())
        defer { engine.close() }

        var chat = LlamaGenerationRequest(messages: [
            LlamaMessage(role: "system", content: "Answer in one short sentence."),
            LlamaMessage(role: "user", content: "What is 2+2?"),
        ])
        chat.maxTokens = 24
        chat.sampling.temperature = 0
        let r = try await engine.generate(chat)   // async form never blocks
        XCTAssertFalse(r.text.isEmpty)
    }

    func testLoadProgressCallbackFires() async throws {
        try requireModel()
        let lock = NSLock()
        var pings = 0
        var c = makeConfig()
        c.loadProgress = { p in
            lock.lock(); pings += 1; lock.unlock()
            return true
        }
        let engine = try await LlamaEngine.open(c)
        engine.close()
        lock.lock(); defer { lock.unlock() }
        XCTAssertGreaterThan(pings, 0, "load progress callback should fire")
    }

    /// Streaming: first event is .started(id), then tokens, then a terminal
    /// event; abort() from another thread stops it with stopReason == .aborted.
    func testStreamStartedAndMidStreamAbort() async throws {
        try requireModel()
        let engine = try await LlamaEngine.open(makeConfig())
        defer { engine.close() }

        var req = LlamaGenerationRequest(prompt: "Write a long, detailed essay about llamas. Do not stop.")
        req.maxTokens = 2000
        req.sampling.temperature = 0.9
        req.sampling.ignoreEos = true

        var startedId: UInt64?
        var tokenCount = 0
        var terminalReason: LlamaStopReason?
        var abortCalled = false
        var streamFailed = false

        for try await ev in engine.generateStream(req) {
            switch ev {
            case .started(let id):
                startedId = id
            case .token:
                tokenCount += 1
                if let sid = startedId, !abortCalled, tokenCount >= 2 {
                    abortCalled = true
                    do { try engine.abort(requestId: sid) }
                    catch { XCTFail("abort threw \(error)") }
                }
            case .done(let r):
                terminalReason = r.stopReason
            case .failed(let e):
                streamFailed = true
                XCTFail("stream failed: \(e)")
            }
        }
        XCTAssertNotNil(startedId)
        XCTAssertGreaterThan(startedId ?? 0, 0, "requestId must be nonzero")
        XCTAssertGreaterThan(tokenCount, 0)
        XCTAssertTrue(abortCalled, "abort was never invoked mid-stream")
        XCTAssertFalse(streamFailed)
        XCTAssertEqual(terminalReason, .aborted)
    }

    /// Streaming a short greedy request terminates normally with .done/.eos.
    func testStreamShortRequestCompletesWithEos() async throws {
        try requireModel()
        let engine = try await LlamaEngine.open(makeConfig())
        defer { engine.close() }

        var req = LlamaGenerationRequest(prompt: "Say exactly: hello")
        req.maxTokens = 16
        req.sampling.temperature = 0

        var text = ""
        var terminalReason: LlamaStopReason?
        for try await ev in engine.generateStream(req) {
            switch ev {
            case .started: break
            case .token(let t): text += t
            case .done(let r): terminalReason = r.stopReason
            case .failed(let e): XCTFail("stream failed \(e)")
            }
        }
        XCTAssertFalse(text.isEmpty)
        XCTAssertEqual(terminalReason, .eos)
    }

    /// Close is idempotent and safe to call after streams completed.
    func testCloseIsIdempotent() async throws {
        try requireModel()
        let engine = try await LlamaEngine.open(makeConfig())
        engine.close()
        engine.close()
        XCTAssertFalse(engine.isOpen)
    }
}

File: llama_mobile-android-SDK/src/androidTest/java/com/llamamobile/LlamaEngineTests.kt
// LlamaEngineTests.kt — v2 Android SDK conformance tests (M5)
//
// Mirrors docs/api-contract-v2.md §8: sync/async generation, single-flight,
// streaming tokens + thread-safe abort, idempotent close. Model-backed tests
// skip when no GGUF is reachable (set LLAMA_MOBILE_TEST_MODEL via a system
// property / adb shell setprop on a real device).

package com.llamamobile

import java.io.File
import java.util.concurrent.CompletableFuture
import java.util.concurrent.TimeUnit
import org.junit.Assert.assertEquals
import org.junit.Assert.assertTrue
import org.junit.Test
import org.junit.runner.RunWith
import androidx.test.ext.junit.runners.AndroidJUnit4
import androidx.test.platform.app.InstrumentationRegistry

@RunWith(AndroidJUnit4::class)
class LlamaEngineTests {

    // Model lives in the test app's own external dir (adb-pushed by the host
    // runner), so no storage permission is needed.
    private fun modelPath(): String? {
        val ctx = InstrumentationRegistry.getInstrumentation().targetContext
        val dir = ctx.getExternalFilesDir(null)
        val own = dir?.let { File(it, "SmolLM-360M-Instruct.Q6_K.gguf") }
        if (own != null && own.exists()) return own.absolutePath
        System.getProperty("LLAMA_MOBILE_TEST_MODEL")?.let { if (File(it).exists()) return it }
        val legacy = listOf(
            "/sdcard/Download/SmolLM-360M-Instruct.Q6_K.gguf",
            "/data/local/tmp/models/SmolLM-360M-Instruct.Q6_K.gguf",
        )
        return legacy.firstOrNull { File(it).exists() }
    }

    private fun modelAvailable(): Boolean = modelPath() != null

    // Embedding model is adb-pushed beside the main model by the runner.
    private fun embeddingModelPath(): String? {
        val ctx = InstrumentationRegistry.getInstrumentation().targetContext
        val dir = ctx.getExternalFilesDir(null)
        val own = dir?.let { File(it, "Qwen3-Embedding-0.6B-Q8_0.gguf") }
        if (own != null && own.exists()) return own.absolutePath
        return listOf(
            "/sdcard/Download/Qwen3-Embedding-0.6B-Q8_0.gguf",
            "/data/local/tmp/models/Qwen3-Embedding-0.6B-Q8_0.gguf",
        ).firstOrNull { File(it).exists() }
    }

    private fun config(): LlamaEngine.Config {
        val c = LlamaEngine.Config()
        c.modelPath = modelPath() ?: ""
        c.nCtx = 2048
        c.engine = 0 // AUTO -> Vulkan on real device
        c.nGpuLayers = 1
        return c
    }

    private fun requireModel() {
        org.junit.Assume.assumeTrue("model not available", modelAvailable())
    }

    @Test
    fun libraryVersionIsV2() {
        val version = LlamaEngine.libraryVersion()
        assertTrue("version string present: $version", version.isNotEmpty())
    }

    @Test
    fun openMissingModelThrows() {
        val c = LlamaEngine.Config()
        c.modelPath = "/nonexistent/model.gguf"
        try {
            LlamaEngine.open(c)
            throw AssertionError("expected open to fail for a missing model")
        } catch (e: LlamaException) {
            assertTrue(
                "got ${e.status}",
                e.status == LlamaStatus.MODEL_LOAD || e.status == LlamaStatus.MODEL_NOT_FOUND ||
                    e.status == LlamaStatus.IO || e.status == LlamaStatus.INVALID_ARGUMENT,
            )
        }
    }

    @Test
    fun syncGenerateProducesText() {
        if (!modelAvailable()) return
        LlamaEngine.open(config()).use { engine ->
            val req = LlamaGenerationRequest(
                prompt = "Complete: the capital of France is",
                maxTokens = 16,
                sampling = LlamaSampling(temperature = 0f, topP = 1.0f, minP = 0f),
            )
            val r = engine.generate(req)
            assertTrue("produced text: '${r.text}'", r.text.isNotEmpty())
            assertEquals(LlamaStopReason.EOS, r.stopReason)
        }
    }

    @Test
    fun asyncGenerateNeverBlocksCaller() {
        if (!modelAvailable()) return
        LlamaEngine.open(config()).use { engine ->
            val f = engine.generateAsync(
                LlamaGenerationRequest(
                    messages = listOf(
                        LlamaChatMessage("system", "Answer in one short sentence."),
                        LlamaChatMessage("user", "What is 2+2?"),
                    ),
                    maxTokens = 24,
                    sampling = LlamaSampling(temperature = 0f),
                ),
            )
            assertTrue(f is CompletableFuture<*>)
            val r = f.get(300, TimeUnit.SECONDS)
            assertTrue(r.text.isNotEmpty())
        }
    }

    @Test
    fun streamingAbortStopsWithAborted() {
        if (!modelAvailable()) return
        LlamaEngine.open(config()).use { engine ->
            val tokens = StringBuilder()
            val started = java.util.concurrent.atomic.AtomicBoolean(false)
            val aborted = java.util.concurrent.atomic.AtomicBoolean(false)
            val f = engine.generateAsync(
                LlamaGenerationRequest(
                    prompt = "Write a long, detailed essay about the history of llamas.",
                    maxTokens = 2000,
                    sampling = LlamaSampling(temperature = 0.9f, ignoreEos = true),
                ),
                tokenCallback = TokenCallback { token ->
                    tokens.append(token)
                    if (started.compareAndSet(false, true)) {
                        // Abort right after the first token arrives.
                        Thread { aborted.set(engine.abort()) }.start()
                    }
                    true
                },
            )
            val r = f.get(300, TimeUnit.SECONDS)
            assertTrue("some tokens arrived", tokens.isNotEmpty())
            if (aborted.get()) {
                // Abort landed mid-generation: it must report ABORTED.
                assertEquals(LlamaStopReason.ABORTED, r.stopReason)
            } else {
                // Generation finished before the abort landed (engine stops on
                // EOS unconditionally) — must still be a clean terminal stop.
                assertTrue(
                    "clean terminal stop, got ${r.stopReason}",
                    r.stopReason == LlamaStopReason.EOS ||
                        r.stopReason == LlamaStopReason.LENGTH,
                )
            }
        }
    }

    /** GPU/Vulkan smoke on a real device: open with GPU layers > 0 and run a
     *  tiny generation. Marks the test skipped when the model is unavailable. */
    @Test
    fun gpuBackendSmoke() {
        if (!modelAvailable()) return
        val c = LlamaEngine.Config()
        c.modelPath = modelPath() ?: ""
        c.nCtx = 1024
        c.engine = 0 // AUTO (device Vulkan when available)
        c.nGpuLayers = 1
        LlamaEngine.open(c).use { engine ->
            val req = LlamaGenerationRequest(
                prompt = "Say hi",
                maxTokens = 8,
                sampling = LlamaSampling(temperature = 0f),
            )
            val r = engine.generate(req, null)
            assertTrue("gpu smoke produced text", r.text.isNotEmpty())
        }
    }

    @Test
    fun closeIsIdempotent() {
        if (!modelAvailable()) return
        val engine = LlamaEngine.open(config())
        engine.close()
        engine.close() // must not throw
    }

    @Test
    fun modelInfoReportsFields() {
        if (!modelAvailable()) return
        LlamaEngine.open(config()).use { engine ->
            val info = engine.modelInfo()
            assertEquals("nCtx should reflect the config", 2048L, info.nCtx.toLong())
            assertTrue("nParams > 0, got ${info.nParams}", info.nParams > 0)
            assertTrue("nEmbd > 0, got ${info.nEmbd}", info.nEmbd > 0)
            assertTrue("description present", info.description.isNotEmpty())
        }
    }

    @Test
    fun tokenizeDetokenizeRoundTrip() {
        if (!modelAvailable()) return
        LlamaEngine.open(config()).use { engine ->
            val tokens = engine.tokenize("Hello from llama mobile v2")
            assertTrue("tokenized to some tokens", tokens.isNotEmpty())
            val round = engine.detokenize(tokens)
            assertTrue("detokenize produced text", round.isNotEmpty())
        }
    }

    @Test
    fun embedProducesVectorsOnEmbeddingModel() {
        val emPath = embeddingModelPath()
        if (emPath == null) return
        val c = LlamaEngine.Config()
        c.modelPath = emPath
        c.nCtx = 512
        c.engine = 0
        c.nGpuLayers = 1
        c.embedding = true
        LlamaEngine.open(c).use { engine ->
            val rows = engine.embed(listOf("hello", "world"))
            assertEquals("one vector per input text", 2L, rows.size.toLong())
            assertTrue("embedding dim > 0", rows[0].isNotEmpty())
            assertEquals(
                "same dim for every row",
                rows[0].size.toLong(),
                rows[1].size.toLong(),
            )
            assertTrue(
                "different texts give different vectors",
                !rows[0].contentEquals(rows[1]),
            )
        }
    }
    private fun extFile(name: String): File? {
        val dir = InstrumentationRegistry.getInstrumentation().targetContext
            .getExternalFilesDir(null) ?: return null
        return File(dir, name)
    }

    /** Vision: SmolVLM + mmproj, image caption (CPU first; GPU variant below). */
    @Test
    fun visionGenerateWithGpu() {
        val model = extFile("SmolVLM-256M-Instruct-Q8_0.gguf") ?: return
        val mmproj = extFile("mmproj-SmolVLM-256M-Instruct-Q8_0.gguf") ?: return
        val img = extFile("image.jpg") ?: return
        if (!model.exists() || !mmproj.exists() || !img.exists()) return
        val c = LlamaEngine.Config()
        c.modelPath = model.absolutePath
        c.nCtx = 2048
        c.chat = true
        c.engine = 1 // CPU first — isolates pipeline from the Vulkan backend
        c.nGpuLayers = 0
        try {
            LlamaEngine.open(c).use { engine ->
                val ok = engine.initMultimodal(mmproj.absolutePath)
                assertTrue("initMultimodal returned false", ok)
                assertTrue("supportsVision true", engine.supportsVision())
                val r = engine.generate(
                    LlamaGenerationRequest(
                        prompt = "Describe this picture in a few words.",
                        mediaPaths = listOf(img.absolutePath),
                        maxTokens = 32,
                        sampling = LlamaSampling(temperature = 0f),
                    ),
                )
                assertTrue("vision caption produced text", r.text.isNotEmpty())
            }
        } catch (t: Throwable) {
            throw AssertionError("vision test failed: " + t, t)
        }
    }

    /** TTS: OuteTTS main + WavTokenizer vocoder -> PCM samples. */
    @Test
    fun ttsSpeakProducesPcm() {
        val model = extFile("OuteTTS-0.2-500M-Q6_K.gguf") ?: return
        val vocoder = extFile("WavTokenizer-Large-75-F16.gguf") ?: return
        if (!model.exists() || !vocoder.exists()) return
        val c = LlamaEngine.Config()
        c.modelPath = model.absolutePath
        c.nCtx = 4096
        c.engine = 1
        c.nGpuLayers = 0
        try {
        LlamaEngine.open(c).use { engine ->
            assertTrue("tts init ok", engine.ttsInit(vocoder.absolutePath))
            assertTrue("tts enabled", engine.ttsEnabled())
            val out = engine.ttsSpeak("Hello from on device speech.")
            assertTrue("pcm samples produced", out.pcm.isNotEmpty())
            assertTrue("usage generated>0", out.usage.generatedTokens > 0)
            assertTrue("tts release ok", engine.ttsRelease())
        }
        } catch (t: Throwable) { throw AssertionError("tts failed: " + t, t) }
    }

    /** Embedding on the GPU (Vulkan). */
    @Test
    fun embedGpu() {
        val model = extFile("Qwen3-Embedding-0.6B-Q8_0.gguf") ?: return
        if (!model.exists()) return
        val c = LlamaEngine.Config()
        c.modelPath = model.absolutePath
        c.nCtx = 512
        c.engine = 0
        c.nGpuLayers = 1
        c.embedding = true
        try {
            LlamaEngine.open(c).use { engine ->
                val rows = engine.embed(listOf("hello", "world"))
                assertTrue("embedding dim > 0", rows.first().isNotEmpty())
            }
        } catch (t: Throwable) { throw AssertionError("embedGpu failed: " + t, t) }
    }

    /** Vision caption on the GPU (Vulkan). */
    @Test
    fun visionGpu() {
        val model = extFile("SmolVLM-256M-Instruct-Q8_0.gguf") ?: return
        val mmproj = extFile("mmproj-SmolVLM-256M-Instruct-Q8_0.gguf") ?: return
        val img = extFile("image.jpg") ?: return
        if (!model.exists() || !mmproj.exists() || !img.exists()) return
        val c = LlamaEngine.Config()
        c.modelPath = model.absolutePath
        c.nCtx = 2048
        c.chat = true
        c.engine = 0
        c.nGpuLayers = 1
        try {
            LlamaEngine.open(c).use { engine ->
                assertTrue("initMultimodal ok", engine.initMultimodal(mmproj.absolutePath))
                val r = engine.generate(
                    LlamaGenerationRequest(
                        prompt = "Describe this picture in a few words.",
                        mediaPaths = listOf(img.absolutePath),
                        maxTokens = 32,
                        sampling = LlamaSampling(temperature = 0f),
                    ),
                )
                assertTrue("vision text produced", r.text.isNotEmpty())
            }
        } catch (t: Throwable) { throw AssertionError("visionGpu failed: " + t, t) }
    }

    /** TTS speak with GPU offload (Vulkan). */
    @Test
    fun ttsGpu() {
        val model = extFile("OuteTTS-0.2-500M-Q6_K.gguf") ?: return
        val vocoder = extFile("WavTokenizer-Large-75-F16.gguf") ?: return
        if (!model.exists() || !vocoder.exists()) return
        val c = LlamaEngine.Config()
        c.modelPath = model.absolutePath
        c.nCtx = 4096
        c.engine = 0
        c.nGpuLayers = 1
        try {
            LlamaEngine.open(c).use { engine ->
                assertTrue("tts init ok", engine.ttsInit(vocoder.absolutePath))
                val out = engine.ttsSpeak("Hello from GPU speech.")
                assertTrue("pcm produced", out.pcm.isNotEmpty())
            }
        } catch (t: Throwable) { throw AssertionError("ttsGpu failed: " + t, t) }
    }

}
